### PR TITLE
cilium: add skb_pull_data to bpf_network to avoid revalidate error

### PR DESCRIPTION
**Pull request by jrfastab**
Depending on the driver when we have a full stack of headers with
tunneled IPsec its possible the headers are not part of the linear
data. In this case we should pull in the data instead of immediately
aborting. If the pull fails we can then abort.

For this patch we only fix the immediate problem call site. We need
to audit the remaining ones. However, veth devices should not have
this issue in general so bpf_lxc.c should be OK. skbs on bpf_netdev
either passed through bpf_network.c, bpf_lxc.c or were routed from
the host networking.

Fixes: 47bd875512770 ("cilium: encrypt, use ipcache to lookup IPsec destination IP")
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8333)

### CODE REVIEW

**Comment by jrfastab** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
Might be nice to fold it into a `revalidate_data_or_pull()` but there's only a couple of users here.

File: bpf/bpf_network.c [hunk 1]
```diff
@@ -44,8 +44,12 @@ static inline int handle_ipv6(struct __sk_buff *skb)
 	bool decrypted;
 
 	decrypted = ((skb->mark & MARK_MAGIC_HOST_MASK) == MARK_MAGIC_DECRYPT);
-	if (!revalidate_data(skb, &data, &data_end, &ip6))
-		return DROP_INVALID;
+	if (!revalidate_data(skb, &data, &data_end, &ip6)) {
+		int err = skb_pull_data(skb, ETH_HLEN + sizeof(struct iphdr));
+
+		if (err || !revalidate_data(skb, &data, &data_end, &ip4))
```
Comment (borkmann) [inline bpf/bpf_network.c]: This looks not correct, above should both be ipv6 hdr, no?

Comment (jrfastab) [reply bpf/bpf_network.c]: Right

Comment (jrfastab) [reply bpf/bpf_network.c]: my test cluster had ipv6 disabled so it didn't catch the compiler error above either. So will fix both and push update.

**Comment by jrfastab** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Pull request merged**